src/core/Type.ml

```diff
@@ -244,7 +244,8 @@ let apply ty0 args0 =
       aux_l ret exp_args args env
     | T.Bind (Binder.ForallTy, _, ty'), arg :: args' ->
       let arg = T.DB.eval env arg in
-      aux ty' args' (DBEnv.push env arg)
+      let env = DBEnv.push env arg in
+      aux (T.DB.eval env ty') args' env
```

Comment:
this is too eager, I think. (it evaluates `ty'` under environment `env` at every recursive call).

I suggest trying adding a case:

```ocaml
    | T.DB _, _ ->
      let ty = T.DB.eval env ty in
      aux ty args env
```

To make things clearer: I use De Bruijn indices for binders (because it goes well with hashconsing of terms and types). An environment `env` is a list of terms accessed by their level. It's effectively a substitution for DB indices, and when you enter a binder you push something at the beginning of the list (thus shifting the indices of all terms already there, which corresponds perfectly to how DB indices are supposed to work).

`T.DB.eval env t` is going to traverse `t` and replace its bound variables by whatever is in `env`.

Here, every time we have `Πα. τ[α]` and an argument `a`, we push `a` at the beginning of the environment, thus binding `α` to it.

The modification I suggest should deal with the case where the toplevel type itself is a bound variable, which is the only one I think is missing.

Reply:
Ok, I think I understood how this works now. You are right. I changed the code using your suggestion.